### PR TITLE
docs: example of csharp example serializer

### LINKED ISSUE #417: Show how to generate C# data models with serializer functionality
## What to solve

Sometimes you want to use the models for data transfers, and while most cases would work out of the box, custom serializer functionality is needed for the advanced cases. 

If you generated the data models based on JSON Schema document and you want the serialized data to validate against the schema, this functionality is REQUIRED.

This can be done including the preset `CSHARP_JSON_SERIALIZER_PRESET`.

Use the following code in the example:
```ts
generator = new CSharpGenerator({ 
  presets: [
    CSHARP_JSON_SERIALIZER_PRESET
  ]
});
```

Duplicate the folder https://github.com/asyncapi/modelina/tree/master/examples/TEMPLATE and rename it to `csharp-generate-serializer`. Update any relevant documentation.

<details>
<summary>Some folks want more specific tasks to complete, or use this if you get stuck - CLICK ME</summary>
<p>

This issue is specially designed for first-time contributors, that want to contribute their very first code. If you are accustomed to how to solve issues in AsyncAPI, you can skip some of the steps.

#### What to know before getting started
Before actually getting started with solving the issue, there are a few things that need to be set up:
- Make sure you familiarize yourself with our [code-of-conduct](https://github.com/asyncapi/.github/blob/master/CODE_OF_CONDUCT.md) and [contributor guidelines](https://github.com/asyncapi/.github/blob/master/CONTRIBUTING.md) 
- We work on forks of repositories and branches in forks, if you are unsure how to set this up you can follow the guide on our [git-workflow setup](https://github.com/asyncapi/.github/blob/master/git-workflow.md).

Feel free to reach out either here in the issue, or on [slack under #tooling](https://www.asyncapi.com/slack-invite) if you have any questions or are stuck, we will gladly help.

## Getting started

This is a task list, that you copy into your PR description to ensure everything is included, this list walks you through each step as to how to solve this issue. It assumes that you have already set up your fork and branch and are ready to commit code.
- [ ] Duplicate the folder `https://github.com/asyncapi/modelina/tree/master/examples/TEMPLATE` and rename it to `csharp-generate-serializer`. This will give you a good base to start out with.
- [ ] Adapt the documentation about the feature and ensure the example are linked correctly.
    - The docs that need to change are the following: [docs/languages/Csharp.md#generate-serializer-and-deserializer-functionality](https://github.com/asyncapi/modelina/blob/master/docs/languages/Csharp.md#generate-serializer-and-deserializer-functionality)
    - The documentation must reference an actual example that is tested and contain no code itself. It is because the docs will inevitably become desynchronized and display incorrect code.
- [ ] Adapt the [examples overview](https://github.com/asyncapi/modelina/tree/master/examples/README.md) to include your new example. 
- [ ] Now it is time to adapt the test code (in examples/csharp-generate-serializer/index.spec.ts) and the implementation (in examples/csharp-generate-serializer/index.ts).
- [ ] Make sure that the test succeeds, by standing in the new example folder and execute `npm run test`.  
- [ ] Make sure that there are no linting errors, by standing in the root folder and execute `npm run lint`. 
- [ ] Create your PR, by selecting your branch that should be merged with the master branch - https://github.com/asyncapi/modelina/compare
- [ ] It is good practice to double-check your PR before requesting a review from someone else, the same way you read an assignment before handing it in.

## Do you have any examples I can check for reference?
We have currently 4 example documentations that you can look for guidance on how to achieve the desired output:
- [How to add logging to the library](https://github.com/asyncapi/modelina/tree/master/docs/advanced.md#add-logging-to-library)
- [Generate models from AsyncAPI documents](https://github.com/asyncapi/modelina/tree/master/docs/usage.md#generate-models-from-asyncapi-documents)
- [Generate models from JSON Schema draft 7 documents](https://github.com/asyncapi/modelina/tree/master/docs/usage.md#generate-models-from-json-schema-draft-7-documents)
- [Generate an interface instead of class](https://github.com/asyncapi/modelina/docs/tree/master/languages/TypeScript.md#generate-an-interface-instead-of-class)

</p>
</details>

### PR DESCRIPTION
**Pull request by ron-debajyoti**
**Description**

- I added the template folder for csharp-generate-serializer and made changes and tests accordingly. Also added a snapshot for the test case.

**Related issue(s)**
Fixes #417

### CODE REVIEW

**Comment by ron-debajyoti** [comment]
Hey @jonaslagoni !

I seem to be getting a SonarCloud error but it's not based on my code tbh.

**Comment by jonaslagoni** [comment]
> Hey @jonaslagoni !
> 
> I seem to be getting a SonarCloud error but it's not based on my code tbh.

Yea, don't worry about it, Monday I will fix the duplicate problems 🙂 So ignore that check for now 😄

**Comment by ron-debajyoti** [comment]
I mistakenly forgot to rebase, resulting in multiple additional commits.

File: examples/csharp-generate-serializer/README.md [hunk 1]
```diff
@@ -0,0 +1,17 @@
+# C# Generate `Generate Serializer`
```
Comment (jonaslagoni) [inline examples/csharp-generate-serializer/README.md:1]:
```suggestion
# C# Generate serializer functions
```

File: examples/csharp-generate-serializer/README.md [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+
+A basic example on how to generate models and want to serialize the data to validate against the schema.
```
Comment (jonaslagoni) [inline examples/csharp-generate-serializer/README.md:3]:
```suggestion
A basic example of how to generate models which includes a way to serialize them into JSON.
```

File: examples/csharp-generate-serializer/index.spec.ts [hunk 3]
```diff
@@ -0,0 +1,15 @@
+const spy = jest.spyOn(global.console, 'log').mockImplementation(() => { return; });
+import {generate} from './index';
+
+describe('Should be able to generate a mode to validate serialized data against its schema', () => {
```
Comment (jonaslagoni) [inline examples/csharp-generate-serializer/index.spec.ts:4]:
```suggestion
describe('Should be able to generate a model with functions to serialize the data model into JSON ', () => {
```

File: examples/README.md [hunk 4]
```diff
@@ -21,6 +21,7 @@ This directory contains a series of self-contained examples that you can use as
 - [TEMPLATE](./TEMPLATE) - A basic template used to create new examples.
 - [java-generate-tostring](./java-generate-tostring) - A basic example that shows how to generate models that overwrite the `toString` method
 - [csharp-generate-equals-and-hashcode](./csharp-generate-equals-and-hashcode) - A basic example on how to generate models that overwrite the `Equal` and `GetHashCode` methods
+- [csharp-generate-serializer](./csharp-generate-serializer) - A basic example on how to generate models that serializes data to validate against the schema
```
Comment (jonaslagoni) [inline examples/README.md:24]:
```suggestion
- [csharp-generate-serializer](./csharp-generate-serializer) - A basic example on how to generate models that include function to serialize the data models to JSON.
```

File: docs/languages/Csharp.md [hunk 5]
```diff
@@ -11,9 +12,13 @@ There are special use-cases that each language supports; this document pertains
 <!-- tocstop -->
 
 ## Generate serializer and deserializer functionality
-TODO
+
+To generate the data models based on JSON Schema document and validate the serialized data against the schema, use the preset `CSHARP_JSON_SERIALIZER_PRESET`
```
Comment (jonaslagoni) [inline docs/languages/Csharp.md:16]:
```suggestion
Sometimes you want to serialize the data models into JSON, in order to do that use the preset `CSHARP_JSON_SERIALIZER_PRESET`
```
It is not only related to JSON schema, even though you are correct that instantiated data models should always validate accurately against the schema 🙂 If the models are created from a schema of course, but this does not necessarily have to be the case.

**Review by jonaslagoni — CHANGES_REQUESTED** [review]
Got a few suggestions in order to re-phrase the description to be more precise 🙂

File: docs/languages/Csharp.md (hunk 5 above)
Comment (ron-debajyoti) [reply docs/languages/Csharp.md:16]: Right, I kind of overlooked this. Sorry !

**Comment by ron-debajyoti** [comment]
I also wanted to point out @jonaslagoni  if do you think a `.nvmrc` file is required for this project since it specifies the node version to be >=14.

**Comment by jonaslagoni** [comment]
> I also wanted to point out @jonaslagoni if do you think a `.nvmrc` file is required for this project since it specifies the node version to be >=14.

Could very well be worth it, but lets discuss it in an issue please 🙂

**Review by jonaslagoni — APPROVED** [review]
LGTM 👍 🎉

**Comment by jonaslagoni** [comment]
@all-contributors please add @ron-debajyoti for doc, example

**Pull request merged**